src/structures/gdt.rs

```diff
@@ -61,57 +118,72 @@ impl GlobalDescriptorTable {
         assert!(MAX > 0, "A GDT cannot have 0 entries");
         assert!(MAX <= (1 << 13), "A GDT can only have at most 2^13 entries");
+
+        // TODO: Replace with inline_const when it's stable.
+        #[allow(clippy::declare_interior_mutable_const)]
+        const NULL: Entry = Entry::new(0);
         Self {
-            table: [0; MAX],
+            table: [NULL; MAX],
             len: 1,
         }
     }
 
     /// Forms a GDT from a slice of `u64`.
     ///
-    /// # Safety
+    /// This method allows for creation of a GDT with malformed or invalid
+    /// entries. However, it is safe because loading a GDT with invalid
+    /// entires doesn't do anything until those entries are used. For example,
+    /// [`CS::set_reg`] and [`load_tss`](crate::instructions::tables::load_tss)
+    /// are both unsafe for this reason.
     ///
-    /// * The user must make sure that the entries are well formed
-    /// * Panics if the provided slice has more than `MAX` entries
+    /// Panics if:
+    /// * the provided slice has more than `MAX` entries
+    /// * the provided slice is empty
+    /// * the first entry is not zero
     #[inline]
-    pub const unsafe fn from_raw_slice(slice: &[u64]) -> Self {
+    pub const fn from_raw_entries(slice: &[u64]) -> Self {
         let len = slice.len();
-        let mut table = [0; MAX];
+        let mut table = Self::empty().table;
         let mut idx = 0;
 
+        assert!(len > 0, "cannot initialize GDT with empty slice");
+        assert!(slice[0] == 0, "first GDT entry must be zero");
         assert!(
             len <= MAX,
             "cannot initialize GDT with slice exceeding the maximum length"
         );
 
         while idx < len {
-            table[idx] = slice[idx];
+            table[idx] = Entry::new(slice[idx]);
             idx += 1;
         }
 
         Self { table, len }
     }
 
-    /// Get a reference to the internal table.
+    /// Get a reference to the internal [`Entry`] table.
     ///
-    /// The resulting slice may contain system descriptors, which span two `u64`s.
+    /// The resulting slice may contain system descriptors, which span two [`Entry`]s.
     #[inline]
-    pub fn as_raw_slice(&self) -> &[u64] {
+    pub fn entries(&self) -> &[Entry] {
         &self.table[..self.len]
     }
 
     /// Adds the given segment descriptor to the GDT, returning the segment selector.
     ///
-    /// Panics if the GDT doesn't have enough free entries to hold the Descriptor.
+    /// Note that depending on the type of the [`Descriptor`] this may add either
+    /// one or two new entries.
+    ///
+    /// Panics if the GDT doesn't have enough free entries.
     #[inline]
     #[cfg_attr(feature = "const_fn", rustversion::attr(all(), const))]
-    pub fn add_entry(&mut self, entry: Descriptor) -> SegmentSelector {
+    pub fn append(&mut self, entry: Descriptor) -> SegmentSelector {
```

Comment:
Small nit: The doc comment should probably also say "Append" instead of "Add" now that the name has been changed.